src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+
+using Internal.NativeFormat;
+using Internal.Runtime;
+using Internal.Runtime.Augments;
+using Internal.Runtime.TypeLoader;
+
+namespace System.Runtime.InteropServices
+{
+    internal static class TypeMapLazyDictionary
+    {
+        public static IReadOnlyDictionary<string, Type> CreateExternalTypeMap(RuntimeType typeMapGroup)
+        {
+            RuntimeTypeHandle typeMapGroupHandle = typeMapGroup.TypeHandle;
+            foreach (TypeManagerHandle module in RuntimeAugments.GetLoadedModules())
+            {
+                if (!TryGetNativeReaderForBlob(module, ReflectionMapBlob.ExternalTypeMap, out NativeReader externalTypeMapReader))
+                {
+                    continue;
+                }
+                NativeParser externalTypeMapParser = new NativeParser(externalTypeMapReader, 0);
+                NativeHashtable externalTypeMapTable = new NativeHashtable(externalTypeMapParser);
+
+                ExternalReferencesTable externalReferences = default;
+                externalReferences.InitializeCommonFixupsTable(module);
+
+                var lookup = externalTypeMapTable.Lookup(typeMapGroupHandle.GetHashCode());
+                NativeParser entryParser;
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    RuntimeTypeHandle foundTypeMapGroup = externalReferences.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                    if (!foundTypeMapGroup.Equals(typeMapGroupHandle))
+                    {
+                        continue;
+                    }
+                    bool isValid = entryParser.GetUnsigned() == 1;
+                    if (!isValid)
+                    {
+                        unsafe
+                        {
+                            delegate*<void> exceptionStub = (delegate*<void>)externalReferences.GetFunctionPointerFromIndex(entryParser.GetUnsigned());
+                            exceptionStub();
+                            Debug.Fail("Expected exception stub to throw an exception.");
+                            return null; // Should never reach here, as the exception stub should throw an exception.
+                        }
+                    }
+
+                    return new ExternalTypeMapDictionary(new NativeHashtable(entryParser), externalReferences);
+                }
+            }
+
+            return new Dictionary<string, Type>(); // Return an empty dictionary if no valid type map is found.
```

Comment:
Random thought: if you only call these APIs with a valid type map group, does ILC trim this path out entirely? If not, would it make sense to just return another instance of `ExternalTypeMapDictionary` with no content, or even just some other dummy type implementing the interface, to avoid constructing that `Dictionary<string, Type>` generic type instantiation? In theory it might save a tiny bit of size in apps that don't use `Dictionary<__Canon, __Canon>` at all? (not sure if it's worth it, just figured I'd mention it).

Reply:
Or, should the compiler produce empty maps as needed and should this throw missing metadata exception instead?